File: perl/validate.pl
#!/usr/bin/perl

use strict;
use warnings;

=head1 NAME

validate.pl -- Market data pipeline output validator

=head1 DESCRIPTION

Reads transformed CSV from transform.pl and validates each record
against the registered schema. Writes a validation report to STDOUT
and exits non-zero if any critical violations are found.

Validation rules:
    CRITICAL: record_id must be present and unique
    CRITICAL: instrument_id must be present and non-empty
    CRITICAL: price must be a positive number with exactly 4 decimal places
    CRITICAL: volume must be a positive integer
    CRITICAL: figi_resolved must be 1 (unresolved FIGIs block the pipeline)
    WARNING:  ts_valid should be 1 (invalid timestamps are logged but not fatal)
    WARNING:  notional should be greater than zero
    INFO:     size_bucket should be one of SMALL / MID / LARGE / BLOCK

Usage:
    perl transform.pl data/transformed_input.csv | perl validate.pl

Exit codes:
    0  -- all records passed, no critical violations
    1  -- one or more critical violations found

=cut

my %VALID_SIZE_BUCKETS = map { $_ => 1 } qw(SMALL MID LARGE BLOCK);

my $in = $ARGV[0] ? do { open my $f, '<', $ARGV[0] or die "$!\n"; $f } : \*STDIN;

my $header_line = <$in>;
chomp $header_line;
my @cols = split(/,/, $header_line);

my %seen_ids;
my @critical_violations;
my @warnings;
my @info;
my $record_count = 0;
my $passed_count = 0;

while (my $line = <$in>) {
    chomp $line;
    next unless length $line;

    $record_count++;
    my %rec;
    @rec{@cols} = split(/,/, $line, scalar @cols);

    my $id  = $rec{record_id} // '';
    my $row = "record_id=$id";
    my $record_ok = 1;

    # CRITICAL: record_id present and unique
    unless (length $id) {
        push @critical_violations, "$row: missing record_id";
        $record_ok = 0;
    }
    if ($seen_ids{$id}++) {
        push @critical_violations, "$row: duplicate record_id";
        $record_ok = 0;
    }

    # CRITICAL: instrument_id present
    unless (defined $rec{instrument_id} && length $rec{instrument_id}) {
        push @critical_violations, "$row: missing instrument_id";
        $record_ok = 0;
    }

    # CRITICAL: price format (positive, 4 decimal places)
    my $price = $rec{price} // '';
    unless ($price =~ /^\d+\.\d{4}$/ && $price + 0 > 0) {
        push @critical_violations,
            "$row: price '$price' is not a positive number with 4 decimal places";
        $record_ok = 0;
    }

    # CRITICAL: volume positive integer
    my $vol = $rec{volume} // '';
    unless ($vol =~ /^\d+$/ && $vol > 0) {
        push @critical_violations, "$row: volume '$vol' is not a positive integer";
        $record_ok = 0;
    }

    # CRITICAL: figi resolved
    my $figi_res = $rec{figi_resolved} // '';
    unless ($figi_res eq '1') {
        push @critical_violations, "$row: figi_resolved=0, FIGI lookup failed";
        $record_ok = 0;
    }

    # WARNING: timestamp valid
    my $ts_valid = $rec{ts_valid} // '';
    unless ($ts_valid eq '1') {
        push @warnings, "$row: ts_valid=0, timestamp format invalid";
    }

    # WARNING: notional positive
    my $notional = $rec{notional} // 0;
    unless ($notional > 0) {
        push @warnings, "$row: notional=$notional is not positive";
    }

    # INFO: size bucket known
    my $bucket = $rec{size_bucket} // '';
    unless ($VALID_SIZE_BUCKETS{$bucket}) {
        push @info, "$row: unknown size_bucket '$bucket'";
    }

    $passed_count++ if $record_ok;
}

close $in if $ARGV[0];

# Write validation report
print "=== Pipeline Validation Report ===\n";
printf "Records processed : %d\n", $record_count;
printf "Records passed    : %d\n", $passed_count;
printf "Critical violations: %d\n", scalar @critical_violations;
printf "Warnings          : %d\n", scalar @warnings;
printf "Info              : %d\n", scalar @info;
print "\n";

if (@critical_violations) {
    print "CRITICAL VIOLATIONS:\n";
    print "  $_\n" for @critical_violations;
    print "\n";
}

if (@warnings) {
    print "WARNINGS:\n";
    print "  $_\n" for @warnings;
    print "\n";
}

if (@info) {
    print "INFO:\n";
    print "  $_\n" for @info;
    print "\n";
}

my $exit_code = @critical_violations ? 1 : 0;
print @critical_violations ? "RESULT: FAIL\n" : "RESULT: PASS\n";
exit $exit_code;

File: lab-workspace/src/validate.py
"""
validate.py -- Market data pipeline output validator

Reads transformed CSV from transform.py and validates each record
against the registered schema. Prints a validation report to stdout
and exits non-zero if critical violations are found.

Usage:
    python src/validate.py data/transformed.csv
"""

import csv
import logging
import sys
from pathlib import Path
from typing import Optional

logging.basicConfig(level=logging.INFO, format="%(levelname)s %(message)s")
logger = logging.getLogger(__name__)

PRICE_PATTERN_4DP = r"^\d+\.\d{4}$"
VALID_SIZE_BUCKETS = {"SMALL", "MID", "LARGE", "BLOCK"}

import re
_price_re = re.compile(r"^\d+\.\d{4}$")


def validate_records(records: list[dict]) -> dict:
    """Validate transformed pipeline records against the registered schema.

    Args:
        records: List of transformed record dicts from transform.py output.

    Returns:
        Dict with keys: record_count, passed_count, critical, warnings, info.
    """
    logger.info(f"Starting validate_records with {len(records)} records")

    seen_ids: set[str] = set()
    critical: list[str] = []
    warnings: list[str] = []
    info: list[str]     = []
    passed = 0

    for r in records:
        record_id = r.get("record_id") or ""
        row = f"record_id={record_id}"
        record_ok = True

        # CRITICAL: record_id present and unique
        if not record_id:
            critical.append(f"{row}: missing record_id")
            record_ok = False
        elif record_id in seen_ids:
            critical.append(f"{row}: duplicate record_id")
            record_ok = False
        seen_ids.add(record_id)

        # CRITICAL: instrument_id present
        if not r.get("instrument_id"):
            critical.append(f"{row}: missing instrument_id")
            record_ok = False

        # CRITICAL: price format (positive, 4 decimal places)
        price_str = r.get("price") or ""
        if not _price_re.match(price_str) or float(price_str) <= 0:
            critical.append(
                f"{row}: price '{price_str}' is not a positive number "
                f"with 4 decimal places"
            )
            record_ok = False

        # CRITICAL: volume positive integer
        vol_str = r.get("volume") or ""
        try:
            vol = int(vol_str)
            if vol <= 0:
                raise ValueError
        except (ValueError, TypeError):
            critical.append(f"{row}: volume '{vol_str}' is not a positive integer")
            record_ok = False

        # CRITICAL: figi resolved
        figi_res = r.get("figi_resolved")
        if str(figi_res) != "1":
            critical.append(f"{row}: figi_resolved=0, FIGI lookup failed")
            record_ok = False

        # WARNING: timestamp valid
        if str(r.get("ts_valid")) != "1":
            warnings.append(f"{row}: ts_valid=0, timestamp format invalid")

        # WARNING: notional positive
        try:
            notional = float(r.get("notional") or 0)
            if notional <= 0:
                warnings.append(f"{row}: notional={notional} is not positive")
        except (ValueError, TypeError):
            warnings.append(f"{row}: notional is not numeric")

        # INFO: size bucket recognised
        bucket = r.get("size_bucket") or ""
        if bucket not in VALID_SIZE_BUCKETS:
            info.append(f"{row}: unknown size_bucket '{bucket}'")

        if record_ok:
            passed += 1

    logger.info(
        f"Completed validate_records: {passed}/{len(records)} passed, "
        f"{len(critical)} critical, {len(warnings)} warnings"
    )

    return {
        "record_count": len(records),
        "passed_count": passed,
        "critical":     critical,
        "warnings":     warnings,
        "info":         info,
    }


def print_report(result: dict) -> int:
    """Print the validation report and return the exit code.

    Args:
        result: Dict returned by validate_records().

    Returns:
        0 if no critical violations, 1 otherwise.
    """
    print("=== Pipeline Validation Report ===")
    print(f"Records processed : {result['record_count']}")
    print(f"Records passed    : {result['passed_count']}")
    print(f"Critical violations: {len(result['critical'])}")
    print(f"Warnings          : {len(result['warnings'])}")
    print(f"Info              : {len(result['info'])}")
    print()

    if result["critical"]:
        print("CRITICAL VIOLATIONS:")
        for v in result["critical"]:
            print(f"  {v}")
        print()

    if result["warnings"]:
        print("WARNINGS:")
        for w in result["warnings"]:
            print(f"  {w}")
        print()

    if result["info"]:
        print("INFO:")
        for i in result["info"]:
            print(f"  {i}")
        print()

    if result["critical"]:
        print("RESULT: FAIL")
        return 1

    print("RESULT: PASS")
    return 0


def main(input_file: str) -> None:
    """Run the validation pipeline stage.

    Args:
        input_file: Path string to the transformed input CSV.
    """
    logger.info(f"Starting validate pipeline for {input_file}")

    import os
        input_path = Path(input_file)
    if not os.path.exists(input_file):
        raise FileNotFoundError(f"Input file not found: {input_path}")

    with input_path.open(newline="", encoding="utf-8") as fh:
        reader = csv.DictReader(fh)
        records = list(reader)

    result   = validate_records(records)
    exit_code = print_report(result)

    logger.info("Validate pipeline complete")
    sys.exit(exit_code)


if __name__ == "__main__":
    if len(sys.argv) != 2:
        print(f"Usage: python {sys.argv[0]} <input_csv>", file=sys.stderr)
        sys.exit(1)
    main(sys.argv[1])
